client/src/templates/Challenges/redux/code-storage-epic.js

```diff
@@ -72,17 +74,44 @@ function clearCodeEpic(action$, state$) {
 }
 
 function saveCodeEpic(action$, state$) {
+  let saveSuccessful;
   return action$.pipe(
-    ofType(types.executeChallenge),
+    ofType(types.executeChallenge, types.saveEditorContent),
     // do not save challenge if code is locked
     filter(() => !isCodeLockedSelector(state$.value)),
     tap(() => {
       const state = state$.value;
       const { id } = challengeMetaSelector(state);
       const files = challengeFilesSelector(state);
-      store.set(id, files);
+      try {
+        store.set(id, files);
+        // Possible fileType values: indexhtml indexjs indexjsx
+        // The files Object always has one of these as the first/only attribute
+        const fileType = Object.keys(files)[0];
+        if (store.get(id)[fileType].contents !== files[fileType].contents) {
+          throw Error('Failed to save to localStorage');
+        }
+        saveSuccessful = true;
+      } catch (e) {
+        saveSuccessful = false;
+      }
     }),
-    ignoreElements()
+    ofType(types.saveEditorContent),
+    switchMap(() =>
+      saveSuccessful
+        ? of(
+            createFlashMessage({
+              type: 'success',
+              message: 'Saved! Your code was saved to localStorage'
+            })
+          )
+        : of(
+            createFlashMessage({
+              type: 'warning',
+              message: 'Oops, your code did not save, localStorage may be full'
+            })
+          )
+    )
```

Comment:
@jakethecoder95 Any way to make this DRY?  The only difference between these two sections is the `type` and `message`?

Reply:
Sure thing, something like this?
createFlashMessage({
  type: saveSuccessful ? 'success' : 'warning',
  message: saveSuccessful 
    ? 'Saved! Your code was saved to localStorage' 
    : Oops, your code did not save, localStorage may be full'
})

Reply:
@jakethecoder95 Something like that.  I don't really like the two ternary statements using `saveSucessful`, but it does seem a bit more concise this way.